### PR TITLE
Creatable Artifacts

### PR DESCRIPTION
**Pull request by natjoe4**
# Summary
Adds new authoring page and ability to create simple Library and Measure draft resources and store them in a trpc+mongo backend.

## New behavior
New Authoring page!!
This page gives you the ability to create new draft Libraries and Measures, as well as enter some custom simple information, like url and identifier

Our trpc backend server is now connected to a mongo database.

## Code changes
 - Created Authoring page and added Authoring button to home screen nav bar
 - Added new routes to support Authoring page and Authoring/[resourceType]/[id] routes
 - Added mongodb connection file
 - Added mongodb basic read, create, and update functions
 - Added trpc router for read, create, and update for Library and Measure resources

# Testing guidance
 - add `MONGODB_URI=mongodb://localhost:27017/draft-repository` to `.env` file
 - `npm run check`
 - Fire up Robo3T
 - `npm run start:all`
 - Click on the Authoring button in the nav bar
 - Select either measure or library from the radio buttons and click `Create new Draft` button
 - Ensure that a proper notification appears, then check RoboMongo `draft-repository` collection to ensure that the resource was created
 - Play around with the url and identifier entry fields, and make sure that submitting the updates works as expected
 - Verify in robo3t that updates are actually going through
 - Try again with another resourceType
 - Take a close look at the setup for mongo and trpc and ensure that is all done as intended.

### CODE REVIEW

File: app/src/pages/_app.tsx [hunk 1]
```diff
@@ -48,6 +61,13 @@ function App({ Component, pageProps }: AppProps) {
               <Navbar.Section grow pt={18}>
                 <ResourceCounts />
               </Navbar.Section>
+              <Navbar.Section px={18} style={{ paddingBottom: 16 }}>
+                <Link href={`/authoring`} style={{ width: '100%' }}>
```
Comment (hossenlopp) [inline app/src/pages/_app.tsx:65]: Is there anything we can do to make this authoring button light up while we are on an authoring page? Perhaps make it that primary blue color or maybe the 'warning' color if that exists in mantine. It would help to know you are in authoring mode.

File: app/src/pages/authoring/index.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      <Radio.Group
+        label="Resource Type"
+        value={resourceType}
+        onChange={val => setResourceType(val as ArtifactResourceType)}
+      >
+        <Group>
+          <Radio label="Measure" value="Measure" />
+          <Radio label="Library" value="Library" />
+        </Group>
+      </Radio.Group>
```
Comment (hossenlopp) [inline app/src/pages/authoring/index.tsx:61]: Thoughts on using a [SegmentedControl](https://mantine.dev/core/segmented-control/) instead of Radio? Might look a little sleeker.

File: app/src/pages/authoring/index.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  const router = useRouter();
+
+  const createResource = () => {
+    const newResource = resourceType === 'Measure' ? { ...MeasureSkeleton } : { ...LibrarySkeleton };
+    newResource.id = uuidv4();
+    draftMutation.mutate({ resourceType, draft: newResource });
+  };
+
+  return (
+    <Group
```
Comment (hossenlopp) [inline app/src/pages/authoring/index.tsx:42]: This page could use a header denoting that it is the authoring. Not sure what makes sense.

File: app/src/components/DraftResourceButtons.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import { Text, Paper, Stack, Center, Loader } from '@mantine/core';
+import { trpc } from '@/util/trpc';
+import SidebarButtonList from './SidebarButtonList';
+
+/**
+ * Component which retrieves all resources and their counts and translates them into buttons and adds
+ * a search button that links to the overall search page
```
Comment (elsaperelli) [inline app/src/components/DraftResourceButtons.tsx:7]: Do we want to add a search button here or just remove this part of the component description?

Comment (elsaperelli) [reply app/src/components/DraftResourceButtons.tsx:7]: I would also specify that this is for draft resources

File: app/src/components/DraftResourceButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    <Stack spacing="xs">
+      {artifactCountError ? (
+        <div
+          style={{
+            padding: '24px'
+          }}
+        >
+          <Paper p="xl" radius="md" withBorder>
+            <Text c="red">
+              Resources could not be displayed due to an error retrieving them from the server: &quot;
```
Comment (elsaperelli) [inline app/src/components/DraftResourceButtons.tsx:34]: Maybe reword this error so it's clear its not about the repository service?

File: app/src/pages/_app.tsx [hunk 6]
```diff
@@ -62,15 +110,30 @@ function App({ Component, pageProps }: AppProps) {
...
+                    </Center>
+                    <Text c="white" weight="bold">
+                      Measure Repository
+                    </Text>
+                  </Group>
+                </Link>
+              </div>
+              <Group position="center" className={classes.navGroup} spacing="xl">
+                <Link href="/">
+                  <Text c={router.pathname === '/' ? 'orange.3' : 'gray.4'}>Repository</Text>
```
Comment (elsaperelli) [inline app/src/pages/_app.tsx:131]: image

Do we want "Repository" to be orange when we are browsing the Measure or Library artifacts that are on the mrs?

File: app/src/pages/authoring/index.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import { v4 as uuidv4 } from 'uuid';
+import { useState } from 'react';
+import { trpc } from '../../util/trpc';
+import { MeasureSkeleton, LibrarySkeleton } from '@/util/authoringFixtures';
+import { useRouter } from 'next/router';
+import { notifications } from '@mantine/notifications';
+import { AlertCircle, CircleCheck } from 'tabler-icons-react';
+import { ArtifactResourceType } from '@/util/types/fhir';
+
+const useStyles = createStyles(theme => ({
```
Comment (elsaperelli) [inline app/src/pages/authoring/index.tsx:11]: `theme` not being used is throwing a lint warning

File: app/src/pages/authoring/index.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  }
+}));
+
+export default function AuthoringPage() {
+  const [resourceType, setResourceType] = useState<ArtifactResourceType>('Measure');
+
+  const ctx = trpc.useContext();
+  const { classes } = useStyles();
+
+  const draftMutation = trpc.draft.createDraft.useMutation({
```
Comment (elsaperelli) [inline app/src/pages/authoring/index.tsx:23]: Overall question here, do we want to immediately create a new Measure or Library when we click create? Or do we want to hold off until the user clicks submit?

File: app/src/server/db/dbOperations.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Updates the resource of the given type with the given id
+ */
+export async function updateDraft(resourceType: ArtifactResourceType, id: string, update: any) {
+  const client = await clientPromise;
+  const collection = client.db().collection(resourceType);
+  return collection.updateOne({ id }, { $set: update });
+}
+
```
Comment (elsaperelli) [inline app/src/server/db/dbOperations.ts:39]: Maybe add a quick comment here for consistency? Unless it's self explanatory enough

File: app/src/server/db/mongodb.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+const uri = process.env.MONGODB_URI;
+const options = {};
+
+let client;
+let clientPromise: Promise<MongoClient>;
+
+if (process.env.NODE_ENV === 'development') {
+  // In development mode, use a global variable so that the value
+  // is preserved across module reloads caused by HMR (Hot Module Replacement).
+  const globalWithMongo = global as typeof globalThis & {
```
Comment (elsaperelli) [inline app/src/server/db/mongodb.ts:16]: Also I am a little confused on what this actually does?

**Review by elsaperelli — CHANGES_REQUESTED** [review]
Good stuff! Just some small comments/questions

File: app/src/server/db/mongodb.ts (hunk 10 above)
Comment (mgramigna) [reply app/src/server/db/mongodb.ts:16]: This ensures that the connection can just be re-used when the hot module replacement (fast hot reloading) happens. It makes sure that when using `npm run dev`, you aren't spamming connections to mongo every time your code updates.

In the actual built production code of the app, this code won't get hit and it will just use one connection since the app won't be continuously restarted

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis

File: app/src/pages/_app.tsx (hunk 6 above)
Comment (mgramigna) [reply app/src/pages/_app.tsx:131]: I went back and forth on this, but yeah I think prob we should just keep repository highlighted when not doing authoring stuff. Maybe it will change in the future

File: app/src/pages/authoring/index.tsx (hunk 8 above)
Comment (mgramigna) [reply app/src/pages/authoring/index.tsx:23]: Unsure for now. Maybe it will become clearer which we should do as we get some more inputs in there for things you can add to the artifact. For now it's probably fine as is

**Review by elsaperelli — APPROVED** [review]

**Review by hossenlopp — APPROVED** [review]
